Menus/PieMenuGUI.cpp

```diff
@@ -1,1416 +1,634 @@
-// Virtual method:  Save
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Saves the complete state of this Slice with a Writer for
-//                  later recreation with Create(Reader &reader);
+		m_FullRadius = 58;
+		m_InnerRadius = 0;
+		m_CursorAngle = 0;
 
-int PieMenuGUI::Slice::Save(Writer &writer) const {
-	Serializable::Save(writer);
+		m_BGBitmap = nullptr;
+	}
 
-	writer.NewPropertyWithValue("Description", m_Description);
-	writer.NewPropertyWithValue("Icon", m_Icon);
-	writer.NewPropertyWithValue("Direction", m_Direction);
-	writer.NewPropertyWithValue("ScriptPath", m_ScriptPath);
-	writer.NewPropertyWithValue("FunctionName", m_FunctionName);
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-	return 0;
-}
+	int PieMenuGUI::Create(Controller *controller, MovableObject *affectedObject) {
+		RTEAssert(controller, "No controller sent to PieMenuGUI on creation!");
 
+		if (!s_CursorBitmap) { s_CursorBitmap = ContentFile("Base.rte/GUIs/Skins/PieCursor.png").GetAsBitmap(); }
 
-//////////////////////////////////////////////////////////////////////////////////////////
-// Method:          Clear
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Clears all the member variables of this PieMenuGUI, effectively
-//                  resetting the members of this abstraction level only.
-
-void PieMenuGUI::Clear()
-{
-    m_pController = 0;
-    m_pActor = 0;
-    m_CenterPos.Reset();
-    m_PieEnabled = DISABLED;
-    m_pHoveredSlice = 0;
-    m_pActivatedSlice = 0;
-    m_PieCommand = PSI_NONE;
-    m_AlreadyActivated = PSI_NONE;
-    m_HoverTimer.Reset();
-    m_BlinkTimer.Reset();
-    m_BlinkMode = NOBLINK;
-
-    m_UpSlice.Reset();
-    m_LeftSlice.Reset();
-    m_DownSlice.Reset();
-    m_RightSlice.Reset();
-
-    m_UpRightSlices.clear();
-    m_UpLeftSlices.clear();
-    m_DownLeftSlices.clear();
-    m_DownRightSlices.clear();
-
-    m_AllSlices.clear();
-    m_SliceGroupCount = 0;
-
-    m_EnablingTimer.Reset();
-    m_Wobbling = false;
-    m_Freeze = false;
-    m_InnerRadius = 0;
-    m_EnabledRadius = 58;
-    m_Thickness = 16;
-    m_pBGBitmap = 0;
-    m_RedrawBG = true;
-    m_CursorAng = 0;
-    m_EnoughInput = false;
-    m_DInputHoldTimer.Reset();
-}
+		if (!m_LargeFont) { m_LargeFont = g_FrameMan.GetLargeFont(); }
+
+		m_MenuController = controller;
+		m_AffectedObject = affectedObject;
+		if (m_AffectedObject) {
+			const Actor *affectedObjectAsActor = static_cast<Actor *>(m_AffectedObject);
+			m_CenterPos = (affectedObjectAsActor ? affectedObjectAsActor->GetCPUPos() : m_AffectedObject->GetPos());
+		}
+
+		if (!m_BGBitmap) {
+			int diameter = (m_FullRadius + std::max(m_BackgroundThickness, m_BackgroundSeparatorSize) + 2) * 2;
+			m_BGBitmap = create_bitmap_ex(8, diameter, diameter);
```

Comment:
You create this BITMAP but don't destroy it.
It leaks whenever you're recreating the pie.